flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/TaskBalancedPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,338 @@
+        private final Map<SlotSharingGroup, List<ExecutionSlotSharingGroup>>
+                paralleledExecutionSlotSharingGroupsMap;
+
+        /**
+         * Record the next round-robin {@link ExecutionSlotSharingGroup} index for {@link
+         * SlotSharingGroup}s.
+         */
+        private final Map<SlotSharingGroup, Integer> slotSharingGroupIndexMap;
+
+        private final Map<ExecutionVertexID, ExecutionSlotSharingGroup>
+                executionSlotSharingGroupMap;
+
+        private final Map<JobVertexID, CoLocationGroup> coLocationGroupMap;
+
+        private final Map<CoLocationConstraint, ExecutionSlotSharingGroup>
+                constraintToExecutionSlotSharingGroupMap;
+
+        private TaskBalancedExecutionSlotSharingGroupBuilder(
+                final SchedulingTopology topology,
+                final Set<SlotSharingGroup> slotSharingGroups,
+                final Set<CoLocationGroup> coLocationGroups) {
+            this.topology = checkNotNull(topology);
+
+            this.coLocationGroupMap = new HashMap<>();
+            for (CoLocationGroup coLocationGroup : coLocationGroups) {
+                for (JobVertexID jobVertexId : coLocationGroup.getVertexIds()) {
+                    coLocationGroupMap.put(jobVertexId, coLocationGroup);
+                }
+            }
+
+            this.constraintToExecutionSlotSharingGroupMap = new HashMap<>();
+            this.paralleledExecutionSlotSharingGroupsMap = new HashMap<>(slotSharingGroups.size());
+            this.slotSharingGroupIndexMap = new HashMap<>(slotSharingGroups.size());
+            this.slotSharingGroupMap = new HashMap<>();
+            this.executionSlotSharingGroupMap = new HashMap<>();
+
+            for (SlotSharingGroup slotSharingGroup : slotSharingGroups) {
+                for (JobVertexID jobVertexId : slotSharingGroup.getJobVertexIds()) {
+                    slotSharingGroupMap.put(jobVertexId, slotSharingGroup);
+                }
+            }
+        }
+
+        private Map<ExecutionVertexID, ExecutionSlotSharingGroup> build() {
+
+            final LinkedHashMap<JobVertexID, List<SchedulingExecutionVertex>> allVertices =
+                    getExecutionVertices(topology);
+
+            initParalleledExecutionSlotSharingGroupsMap(allVertices);
+
+            // Loop on job vertices
+            for (Map.Entry<JobVertexID, List<SchedulingExecutionVertex>> executionVertexInfos :
+                    allVertices.entrySet()) {
+
+                JobVertexID jobVertexID = executionVertexInfos.getKey();
+                List<SchedulingExecutionVertex> executionVertices = executionVertexInfos.getValue();
+                final int jobVertexParallel = executionVertices.size();
+                final SlotSharingGroup slotSharingGroup = slotSharingGroupMap.get(jobVertexID);
+                final int startIndex = getSlotRoundRobinIndex(jobVertexParallel, slotSharingGroup);
```

Comment:
This line should be moved into `allocateNonCoLocatedVertices` method, because it's only used there.